protocol/streaming/grpc/grpc_streaming_manager.go

```diff
@@ -9,30 +13,125 @@ var _ types.GrpcStreamingManager = (*GrpcStreamingManagerImpl)(nil)
 
 // GrpcStreamingManagerImpl is an implementation for managing gRPC streaming subscriptions.
 type GrpcStreamingManagerImpl struct {
+	sync.Mutex
+
+	// orderbookSubscriptions maps subscription IDs to their respective orderbook subscriptions.
+	orderbookSubscriptions map[uint32]*OrderbookSubscription
+	nextId                 uint32
+}
+
+// OrderbookSubscription represents a active subscription to the orderbook updates stream.
+type OrderbookSubscription struct {
+	clobPairIds []uint32
+	srv         clobtypes.Query_StreamOrderbookUpdatesServer
+	finished    chan bool
 }
 
 func NewGrpcStreamingManager() *GrpcStreamingManagerImpl {
-	return &GrpcStreamingManagerImpl{}
+	return &GrpcStreamingManagerImpl{
+		orderbookSubscriptions: make(map[uint32]*OrderbookSubscription),
+	}
 }
 
 func (sm *GrpcStreamingManagerImpl) Enabled() bool {
 	return true
 }
 
 // Subscribe subscribes to the orderbook updates stream.
+// This function returns a channel that is used to signal termination when an error occurs.
 func (sm *GrpcStreamingManagerImpl) Subscribe(
 	req clobtypes.StreamOrderbookUpdatesRequest,
 	srv clobtypes.Query_StreamOrderbookUpdatesServer,
 ) (
 	finished chan bool,
 	err error,
 ) {
-	return nil, nil
+	clobPairIds := req.GetClobPairId()
+
+	// Perform some basic validation on the request.
+	if len(clobPairIds) == 0 {
+		return nil, clobtypes.ErrInvalidGrpcStreamingRequest
+	}
+
+	finished = make(chan bool)
+	subscription := &OrderbookSubscription{
+		clobPairIds: clobPairIds,
+		srv:         srv,
+		finished:    finished,
+	}
+
+	sm.Lock()
+	defer sm.Unlock()
+
+	sm.orderbookSubscriptions[sm.nextId] = subscription
+	sm.nextId++
+
+	return finished, nil
 }
 
 // SendOrderbookUpdates groups updates by their clob pair ids and
 // sends messages to the subscribers.
 func (sm *GrpcStreamingManagerImpl) SendOrderbookUpdates(
-	updates *clobtypes.OffchainUpdates,
+	offchainUpdates *clobtypes.OffchainUpdates,
 ) {
+	// Group updates by clob pair id.
+	updates := make(map[uint32]*clobtypes.OffchainUpdates)
+	for _, message := range offchainUpdates.Messages {
+		clobPairId := message.OrderId.ClobPairId
+		if _, ok := updates[clobPairId]; !ok {
+			updates[clobPairId] = clobtypes.NewOffchainUpdates()
+		}
+		updates[clobPairId].Messages = append(updates[clobPairId].Messages, message)
+	}
```

Comment:
Doesn't have to be now, but do we want to condense the messages somehow? We have this condensing function for indexer, but i don't think it works in this case. these events might be condensed differently.
```
func (om *OffchainUpdates) CondenseMessagesForReplay() {
```

Reply:
I think this gets called before `SendOffchainMessages` is called? i.e. the message is already condensed

Reply:
It gets called during Replay Operations but i feel it might not be used half the time due to this code:
```
	if replayUpdates != nil {
		offchainUpdates = replayUpdates
	}
```
I think it should be fine, unless you can think of any edge case where condensing messages drops anything vital for orderbook updates

Reply:
yeah should be fine to send for example (place + remove) for example